src/zkevm_specs/evm/instruction.py

```diff
@@ -368,6 +372,38 @@ def range_check(self, value: Expression, n_bytes: int) -> bytes:
         except OverflowError:
             raise ConstraintUnsatFailure(f"Value {value} has too many bytes to fit {n_bytes} bytes")
 
+    # Return a tuple of `abs(x)` and `x_is_neg`. For a special case when
+    # `x = -(1 << 255)`, this function returns the same value of `-(1 << 255)`,
+    # since it is signed overflow.
+    def abs_word(self, x: RLC) -> Tuple[RLC, FQ]:
+        is_neg = self.word_is_neg(x)
+
+        # Generate the witness `x_abs`.
+        x_abs = x if is_neg == 0 else self.rlc_encode((1 << 256) - x.int_value, 32)
+
+        # Check sign overflow if `x == -(1 << 255)`, since `abs(-(1 << 255))`
+        # should be equal to `-(1 << 255)`.
+        is_overflow = self.is_equal(FQ(x.le_bytes[31]), FQ(128)) * self.is_zero(
+            self.sum(x.le_bytes[:31])
+        )
+
+        x_abs_lo, x_abs_hi = self.word_to_lo_hi(x_abs)
+        x_lo, x_hi = self.word_to_lo_hi(x)
+
+        # Constrain `x_abs_lo == x_lo` and `x_abs_hi == x_hi` if non negative.
+        self.constrain_zero((x_abs_lo - x_lo) * (1 - is_neg))
+        self.constrain_zero((x_abs_hi - x_hi) * (1 - is_neg))
+
+        # When negative, constrain `x_abs_lo + x_lo == 2^128` if x is not
+        # overflow, otherwise `x_abs_lo + x_lo == 0`.
+        self.constrain_zero((x_abs_lo + x_lo - (1 << 128) * (1 - is_overflow)) * is_neg)
+
+        # When negative, constrain `x_abs_hi + x_hi == 2^128 - 1` if x is not
+        # overflow, otherwise `x_abs_hi + x_hi == 2^128`.
+        self.constrain_zero((x_abs_hi + x_hi + 1 - is_overflow - (1 << 128)) * is_neg)
```

Comment:
Yes. I see. It seems that the previous constraints are not right for all cases. I will update the constraints. Thanks a lot.

Reply:
Hi @adria0 and @icemelon ,
After some investigation, I think I should fix to constrain `x + x_abs == 1 << 256` when x is negative. Even if `x = -(1 << 255)` (0x8000...) that is signed overflow, and `abs(1 << 255) = -(1 << 255)` (0x8000...).

So I referenced the code of [instruction.add_words](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/src/zkevm_specs/evm/instruction.py#L374) and fixed the constraints as below:

```
carry_lo, remainder_lo = divmod(x_lo.n + x_abs_lo.n, 1 << 128)
carry_hi, remainder_hi = divmod(x_hi.n + x_abs_hi.n + carry_lo, 1 << 128)

# When `is_neg`, constrain both low and high remainders are zero, and
# `carry_hi == 1`. Since the final result is `1 << 256`.
self.constrain_zero(FQ(remainder_lo + remainder_hi) * is_neg)
self.constrain_zero(FQ(1 - carry_hi) * is_neg)
```

Could you help review again if it is correct when you have time? Thanks a lot.

Reply:
Yes. I see. I try to reference the code of [AddWordsGadget](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/util/math_gadget.rs#L186) in zkevm_circuit, and add the below constraints.

```
# Contrain `sum([x_lo, x_abs_lo]) == sum_lo + carry_lo * 2^128`.
self.constrain_zero(FQ(sum_lo) + FQ(carry_lo) * FQ(1 << 128) - self.sum([x_lo, x_abs_lo]))

# Contrain `sum([x_hi, x_abs_hi]) + carry_lo == sum_hi + carry_hi * 2^128`.
self.constrain_zero(
    FQ(sum_hi) + FQ(carry_hi) * FQ(1 << 128) - FQ(carry_lo) - self.sum([x_hi, x_abs_hi])
)
```